### PR TITLE
SV-Comp changes

### PR DESCRIPTION
**Pull request by AdamZsofi**
This Branch adds features to Gazer, which are required for SV-Comp
### Added features
- Witness generation: SV-Comp requires a standard output format instead of trace.
Using the witness and hash flags, gazer can now generate a violation witness based on trace and a hardcoded correctness witness [more about witnesses](https://github.com/sosy-lab/sv-witnesses).
- gazer_starter.py: A script, handling combined verification using gazer-bmc, gazer-theta and the generated test harnesses (the latter is for filtering incorrect false results).
**Usage:** `./gazer_starter.py --output output_directory task.c`
*(The output directory is for the witness and the test harness)*
*(`./gazer_start.py --version` is also implemented)*

### Minor changes
- In gazer-theta Theta is called with a bigger max heap and stack size. (Right now this is hardcoded, I'll open an issue about it for later)
- The output format of version has been changed to be simpler and more similar to Theta.
Now it looks like this: `Gazer v1.1.1 LLVM v9.0.0`

The version number has been bumped up to v1.1.1 as well.

### CODE REVIEW

File: src/Witness/CorrectnessWitnessWriter.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+
+#include "gazer/Core/LiteralExpr.h"
+#include "gazer/Witness/WitnessWriter.h"
+#include <llvm/ADT/SmallString.h>
+
+#include <fstream>
+#include <time.h>
```
Comment (sallaigy) [inline src/Witness/CorrectnessWitnessWriter.cpp:24]:
```suggestion
#include <ctime>
```

File: src/Witness/CorrectnessWitnessWriter.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    time ( &rawtime );
+    ptm = gmtime( &rawtime ); // UTC timestamp
+    std::stringstream timestamp;
+    timestamp << ptm->tm_year+1900 << "-";
+    timestamp << std::setfill('0') << std::setw(2);
+    timestamp << ptm->tm_mon+1 << "-"
+              << std::setfill('0') << std::setw(2) << ptm->tm_mday << "T" 
+              << std::setfill('0') << std::setw(2) << ptm->tm_hour%24 << ":" 
+              << std::setfill('0') << std::setw(2) << ptm->tm_min << ":" 
+              << std::setfill('0') << std::setw(2) << ptm->tm_sec;
```
Comment (sallaigy) [inline src/Witness/CorrectnessWitnessWriter.cpp:42]: I am not very familiar with date handling in C++, but wouldn't something like `std::put_time` work here?

File: src/Witness/ViolationWitnessWriter.cpp [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    //                 break;
+    //         }
+    //     }
+    //     else {
+    //         expr->print(mOS);
+    //     }
+
+    //     mOS << ";</data>\n";
+    //     // writeLocation(event.getLocation());
+    //     closeEdge();
```
Comment (sallaigy) [inline src/Witness/ViolationWitnessWriter.cpp:135]:
```suggestion
```

File: include/gazer/Witness/WitnessWriter.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+private:
+    void visit(AssignTraceEvent& event) override;
+    void visit(FunctionEntryEvent& event) override;
+    void visit(FunctionReturnEvent& event) override;
+    void visit(FunctionCallEvent& event) override;
+    void visit(UndefinedBehaviorEvent& event) override;
+
+    unsigned int nodeCounter = 0; // the values is always the id of the next node, that hasn't been created yet
+    bool inProgress = false; // true, if witness is initialized, but not closed
+    Trace& trace;
```
Comment (sallaigy) [inline include/gazer/Witness/WitnessWriter.h:72]: Do we need this field? The `write` method will receive the trace anyways.

File: src/Witness/ViolationWitnessWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    // }
+}
+
+// Megj. amúgy ilyen eventre még nem láttam példát a kimenetben
```
Comment (sallaigy) [inline src/Witness/ViolationWitnessWriter.cpp:139]:
```suggestion
```

**Review by sallaigy — CHANGES_REQUESTED** [review]
Very nice job! I have left some comments, but most of them are code style fixes.

**Review by hajduakos — CHANGES_REQUESTED** [review]
Nice job. I have left some comments, mostly stylistic ones.

File: gazer_starter.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    if(timeout!=0):
+        try:
+            output = run_with_timeout(command, timeout)
+            print_if_not_empty(output)
+            if(output != None):
+                parse_output(output, task_with_path)
+            print_line()
+        except subprocess.CalledProcessError as err:
+            print_if_not_empty(err.output)
+        except subprocess.TimeoutExpired as err:
```
[the remark this answers is not in the document]
Comment (hajduakos) [reply gazer_starter.py:184]: (This also applies to other places where exceptions are handled, e.g. `run_with_timeout` functions)

[the remark this answers is not in the document]
Comment (AdamZsofi) [reply gazer_starter.py:198]: I added some generic excepts to the appropriate places, where we can handle them.

File: include/gazer/Witness/WitnessWriter.h (hunk 4 above)
Comment (AdamZsofi) [reply include/gazer/Witness/WitnessWriter.h:72]: No, we only needed it earlier, I'll delete it.

File: src/Witness/CorrectnessWitnessWriter.cpp (hunk 2 above)
Comment (AdamZsofi) [reply src/Witness/CorrectnessWitnessWriter.cpp:42]: You are right, I could have thought, that the standard library should have something for this. `timestamp << std::put_time(ptm, "%FT%T");` handles ISO 8601 time perfectly for us. I'll update it.

File: src/Witness/ViolationWitnessWriter.cpp [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    mOS << "<data key=\"creationtime\">" << timestamp.str() << "</data>\n";
+    mOS << "<data key=\"programfile\">" << src_filename << "</data>\n";
+    createNode(); // entry node
+    inProgress = true;
+}
+
+void ViolationWitnessWriter::closeWitness() {
+    createNode(true);
+    openEdge();
+    // ide lenne jó esetleg beírni a violation linet TODO
```
[the remark this answers is not in the document]
Comment (AdamZsofi) [reply src/Witness/ViolationWitnessWriter.cpp:85]: That TODO is deprecated, I just forgot to delete it.

File: src/Witness/ViolationWitnessWriter.cpp [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    closeEdge();
+    mOS << "</graph>\n</graphml>";
+    inProgress = false;
+}
+
+void ViolationWitnessWriter::visit(AssignTraceEvent& event) {
+    // assert(inProgress && "Witness should be initialized before write and closed after write");
```
[the remark this answers is not in the document]
Comment (AdamZsofi) [reply src/Witness/ViolationWitnessWriter.cpp:92]: Yes, I'll remove them. These functions should be implemented in the future as enhancements of witness generation, but for that we'll have to resolve some complicated issues, so they'll remain empty for now.

**Review by hajduakos — APPROVED** [review]

**Pull request merged**